packages/ketcher-react/src/script/editor/tool/atom.ts

```diff
@@ -65,174 +65,199 @@ class AtomTool {
-
-    dragCtx.action = fromBondAddition(
-      rnd.ctab,
-      this.#bondProps,
-      dragCtx.item.id,
-      Object.assign({}, this.atomProps),
-      undefined,
-      newAtomPos
-    )[0]
-    this.editor.update(dragCtx.action, true)
+    if (atomId !== undefined) {
+      const atom = molecule.atoms.get(atomId)
+      let angle = utils.calcAngle(atom?.pp, rnd.page2obj(event))
+      if (!event.ctrlKey) angle = utils.fracAngle(angle, null)
+      const degrees = utils.degrees(angle)
+      editor.event.message.dispatch({ info: degrees + 'º' })
+      const newAtomPos = utils.calcNewAtomPos(
+        atom?.pp,
+        rnd.page2obj(event),
+        event.ctrlKey
+      )
+
+      if (dragCtx.action) {
+        dragCtx.action.perform(reStruct)
+      }
+
+      dragCtx.action = fromBondAddition(
+        rnd.ctab,
+        this.#bondProps,
+        atomId,
+        Object.assign({}, atomProps),
+        undefined,
+        newAtomPos
+      )[0]
+
+      editor.update(dragCtx.action, true)
+    }
   }
 
   mouseup(event) {
-    const struct = this.editor.render.ctab
-    const molecule = struct.molecule
-    const functionalGroups = molecule.functionalGroups
-    const ci = this.editor.findItem(event, ['atoms', 'bonds'])
-    const atomResult: Array<number> = []
-    const result: Array<number> = []
-    if (ci && functionalGroups && ci.map === 'atoms') {
-      const atomId = FunctionalGroup.atomsInFunctionalGroup(
-        functionalGroups,
-        ci.id
-      )
-      if (atomId !== null) atomResult.push(atomId)
-    }
-    if (atomResult.length > 0) {
-      for (const id of atomResult) {
-        const fgId = FunctionalGroup.findFunctionalGroupByAtom(
-          functionalGroups,
-          id
-        )
-        if (fgId !== null && !result.includes(fgId)) {
-          result.push(fgId)
+    const {
+      dragCtx,
+      atomProps,
+      editor,
+      editor: {
+        render: rnd,
+        render: {
+          ctab: reStruct,
+          ctab: {
+            molecule,
+            molecule: { functionalGroups }
+          }
         }
       }
-      this.editor.event.removeFG.dispatch({ fgIds: result })
-      return
-    }
+    } = this
 
-    if (this.dragCtx) {
-      const dragCtx = this.dragCtx
-      const rnd = this.editor.render
+    const ci = editor.findItem(event, ['atoms', 'bonds', 'functionalGroups'])
+    const action = new Action()
 
-      this.editor.update(
-        dragCtx.action ||
-          (dragCtx.item
-            ? fromAtomsAttrs(rnd.ctab, dragCtx.item.id, this.atomProps, true)
-            : fromAtomAddition(rnd.ctab, rnd.page2obj(event), this.atomProps))
+    if (!dragCtx.item && !ci) {
+      action.mergeWith(
+        fromAtomAddition(reStruct, rnd.page2obj(event), atomProps)
       )
+    } else if (dragCtx.item && ci) {
+      if (dragCtx.item.id === ci.id) {
+        if (
+          ci.map === 'functionalGroups' &&
+          FunctionalGroup.isContractedFunctionalGroup(ci.id, functionalGroups)
```

Comment:
that's fun that I have the same logic in my PR  :) I think we need to extract this into a separate function and reuse it. 
https://github.com/epam/ketcher/pull/2580/files#diff-ee62e3d2a7eb8e9230e2d13e3cfe44e4ea9ea805a61f971355d38aab66dda856R62